proc.c

```diff
@@ -1273,6 +1273,34 @@ rb_proc_get_iseq(VALUE self, int *is_proc)
     return NULL;
 }
 
+/* call-seq:
+ *   prc == other -> true or false
+ *   prc.eql?(other) -> true or false
+ *
+ * Two proc are the same if, and only if, they were created from the same code block.
+ *
+ *   def return_block(&block)
+ *     block
+ *   end
+ *
+ *   def pass_block_twice(&block)
+ *     [return_block(&block), return_block(&block)]
+ *   end
+ *
+ *   block1, block2 = pass_block_twice { puts 'test' }
+ *   # blocks were turned into Proc objects late, only in return_block,
+ *   # so it is two different objects
+ *   block1.object_id == block2.object_id
```

Comment:
Are you sure this is assured? In all implementations? Looks like a bug, or at least a missed opportunity for an optimization and I don't think we should encourage people counting on this behavior.
How about wording it differently, like "comparing by identity may or may not return the same result" and no example?

Reply:
> Are you sure this is assured? In all implementations?

Umm... I believe that's the very cause of the addion of `==` to `Proc`: before Ruby 2.5 they were the same object (because were converted to Proc on passing to `pass_block_twice`, and then the SAME proc was passed to `return_block`, and 2.5 optimized it to NOT be instantiated into an object if the block is just passed further, so instantiation happens two different times in `return_block`).

But yeah, I didn't think that it might be different in other impls, I'll apply your suggestion.

Reply:
Right. But further optimization could track if bloc is instantiated in the first call to `return_block` and if so, reuse that... I'm pretty sure it's better to express the fact that one should rely on `==/eql?` and simply not use the result of `equal?`